apps/src/lab2/responseValidators.ts

```diff
@@ -46,6 +46,21 @@ const PythonSourceResponseValidator: ResponseValidator<
   return sourceValidatorHelper(response, pythonValidator);
 };
 
+// Validator for labs that use JSON sources
+const JsonSourceResponseValidator: ResponseValidator<
+  ProjectSources
+> = response => {
+  const jsonValidator = (responseToValidate: Record<string, unknown>) => {
+    try {
+      JSON.parse(responseToValidate.source as string);
+    } catch (e) {
+      throw new ValidationError('Error parsing JSON: ' + e);
```

Comment:
Looks like you're following the existing pattern, but for my understanding, where does this get caught? I can't seem to trace it properly 🤦

Reply:
No worries this code is a bit tricky to follow 🙂 this `ValidationError` is eventually caught by ProjectManager when it attempts to load sources for the project: https://github.com/code-dot-org/code-dot-org/blob/7b86d9cbbe8bfff2eac03f2ce9adacbd9ceef61f/apps/src/lab2/projects/ProjectManager.ts#L88-L91

It takes care of defaulting to empty sources and thereby not passing the invalid sources to labs.

Reply:
Thanks for the explanation. Where are sources assigned `{}`?

Reply:
Oh, they're just never assigned (they stay undefined from https://github.com/code-dot-org/code-dot-org/blob/806ab49f82ad35393b2873d20f8c6cac956a0264/apps/src/lab2/projects/ProjectManager.ts#L79